tests/integration_tests/functional/test_drive_vhost_user.py

```diff
@@ -60,3 +113,182 @@ def test_vhost_user_block(microvm_factory, guest_kernel, rootfs_ubuntu_22):
     # Verify if guest can run commands.
     exit_code, _, _ = vm.ssh.run("ls")
     assert exit_code == 0
+
+    # Prepare the input for doing the assertion
+    assert_dict = {}
+    # Keep an array of strings specifying the location where some string
+    # from the output is located.
+    # 1-0 means line 1, column 0.
+    keys_array = ["1-0", "1-6"]
+    # Keep a dictionary where the keys are the location and the values
+    # represent the input to assert against.
+    assert_dict[keys_array[0]] = "ro"
+    assert_dict[keys_array[1]] = "/dev/vda"
+    _check_drives(vm, assert_dict, keys_array)
+
+
+def test_device_ordering(microvm_factory, guest_kernel, rootfs_ubuntu_22):
+    """
+    Verify device ordering.
+
+    The root device should correspond to /dev/vda in the guest and
+    the order of the other devices should match their configuration order.
+    """
+
+    VHOST_USER_SOCKET = "/vub.socket"
+
+    vm = microvm_factory.build(guest_kernel, None, monitor_memory=False)
+
+    # Converting path from tmpfs ("./srv/..") to local
+    # path on the host ("../build/..")
+    rootfs_path = utils.to_local_dir_path(str(rootfs_ubuntu_22))
+    # Launching vhost-user-block backend
+    _backend = spawn_vhost_user_backend(vm, rootfs_path, VHOST_USER_SOCKET)
+
+    # We need to setup ssh keys manually because we did not specify rootfs
+    # in microvm_factory.build method
+    ssh_key = rootfs_ubuntu_22.with_suffix(".id_rsa")
+    vm.ssh_key = ssh_key
+    vm.spawn()
+    vm.basic_config(add_root_device=False)
+    vm.add_net_iface()
+
+    # Adding first block device.
+    fs1 = drive_tools.FilesystemFile(os.path.join(vm.fsfiles, "scratch1"), size=128)
+    vm.add_drive("scratch1", fs1.path)
```

Comment:
can we generalise this test to:
1. 2 x virtio + 1 vhost-user where one of the virtio drives is rootfs
2. 1 x virtio + 2 vhost-user where one of the vhost-user drives is rootfs ?

Reply:
What will be the difference? Here we just need to test the ordering of vhost-user-block devices. I can add another vhost-user-block device as 4th device to test that it remains last device after reordering.

Reply:
added 4th device to the test to verify that it remains last after reordering

Reply:
yes, what we would like to verify is a) we preserve ordering of non-rootfs drives inside virtio and vhost-user classes b) we preserve ordering across virtio and vhost-user devices c) rootfs always takes priority (be it virtio or vhost-user)

Reply:
I think this one test verifies a), b) and c) (taking into consideration that there is a similar test for virtio-block only)